launch/launch/launch_service.py

```diff
@@ -338,28 +338,34 @@ def _on_exception(loop, context):
                         ret = await self._shutdown(reason='idle', due_to_sigint=False)
                         assert ret is None, ret
                         continue
+
+                    # Stop running if we're shutting down and there's no more work
+                    if self.__shutting_down and is_idle:
+                        break
+
+                    # Collect futures to wait on
+                    entity_futures = [pair[1] for pair in self._entity_future_pairs]
+                    entity_futures.extend(self.__context._completion_futures)
+
+                    # Create a task to process any events in the queue
+                    # and add it to the list of awaitables
                     process_one_event_task = this_loop.create_task(self._process_one_event())
-                    if self.__shutting_down:
-                        # If shutting down and idle then we're done.
-                        if is_idle:
+                    entity_futures.append(process_one_event_task)
+
+                    # Wait on events and futures until an event is processed or a timeout occurs
+                    while True:
+                        done, pending = await asyncio.wait(
+                            entity_futures,
+                            timeout=1.0,
+                            return_when=asyncio.FIRST_COMPLETED
+                        )
+                        if process_one_event_task in done:
+                            break
+                        elif len(done) == 0:
                             process_one_event_task.cancel()
                             break
-                        entity_futures = [pair[1] for pair in self._entity_future_pairs]
-                        entity_futures.append(process_one_event_task)
-                        entity_futures.extend(self.__context._completion_futures)
-                        done = set()  # type: Set[asyncio.Future]
-                        while not done:
```

Comment:
before, we were waiting on all completion futures to be completed even when shutdown was called.
Now we aren't.

Reply:
TBH, this condition was confusing to me. `while not done` will be true as soon as a single awaitable is done (the set is not empty), not all of them.

In this PR, we actually do wait for all futures, because `is_idle` will be `False` as long as there is still futures not done.

Reply:
> TBH, this condition was confusing to me. while not done will be true as soon as a single awaitable is done (the set is not empty), not all of them.

Yes, this is confusing because we're always waiting on the same future, it's not that we delete from the collection the ones that were already done ....

I really don't know what was the intention of the previous code